### DISCUSSION TITLE
How to create a zero size view of a Buffer with bind groups

### DISCUSSION THREAD

**Discussion by Pauan** [gfx-rs/wgpu#4090] [Q&A] [answered]
I am creating a utility that copies a `Vec<T>` into a `Buffer`, so that way I can easily send data to the GPU.

When the `Vec<T>` increases in size, I want to automatically resize the `Buffer`.

The way that I'm doing this is simple: I create a `Buffer` that is the same size as `vec.capacity()`:

```rust
let size = vec.capacity() * std::mem::size_of::<T>();

let buffer = device.create_buffer(&BufferDescriptor {
    label: ...
    size: size,
    usage: ...
    mapped_at_creation: true,
});
```

And whenever the `Vec` capacity changes, I destroy the old `Buffer` and create a new `Buffer` with the new capacity.

I use the `capacity` instead of the `len` because I don't want to constantly recreate the `Buffer`, I want to amortize the cost of recreating the `Buffer`.

But the capacity of the `Vec` is larger than the len of the `Vec`, so I need to create a slice of the `Buffer` which only contains the len of the `Vec`.

I can do that easily with `buffer.slice`:

```rust
// Calculate byte size of the Vec
let size = vec.len() * std::mem::size_of::<T>();

buffer.slice(..size)
```

However that only works for vertex buffers and index buffers. When I try to create a bind group, I run into an issue:

```rust
// Calculate byte size of the Vec
let size = vec.len() * std::mem::size_of<T>;

let bind_group = device.create_bind_group(&BindGroupDescriptor {
    label: None,
    layout: ...
    entries: &[
        BindGroupEntry {
            binding: 0,
            resource: BindingResource::Buffer(BufferBinding {
                buffer: &buffer,
                offset: 0,
                size: NonZeroU64::new(size),
            }),
        },
    ],
});
```

In this case, if `vec.len()` is `0` then this fails, because the `size` of a `BufferBinding` cannot be `0`.

**Comment by cwfitzgerald** [comment, marked as answer]
The simple answer is that you can't - you need to keep the buffer no smaller than the minimum binding size of the binding (which in the case of an array, is the size of a single element)

**Reply by Pauan** [reply]
So you're saying that it's completely impossible to have a uniform buffer which is empty?

**Reply by cwfitzgerald** [reply]
You can have a buffer which is empty, you just can't construct a shader which it is valid to bind to.

**Reply by Pauan** [reply]
Alright, I'll have to figure out a different solution then, thanks for your help!
